File: frontend/src/types/operations.ts
export type OperationalHealth = 'HEALTHY' | 'ATTENTION' | 'DEGRADED'
export type OperationalSeverity = 'INFO' | 'WARNING' | 'CRITICAL'
export type OperationalIncidentStatus = 'OPEN' | 'ACKNOWLEDGED' | 'RESOLVED'

export interface OperationalIncidentEvent {
  id: string
  event_type: 'OPENED' | 'UPDATED' | 'ACKNOWLEDGED' | 'RESOLVED'
  from_status: OperationalIncidentStatus | null
  to_status: OperationalIncidentStatus
  source: string
  reason: string
  evidence: Record<string, unknown>
  created_at: string
}

export interface OperationalIncident {
  id: string
  incident_type: string
  severity: OperationalSeverity
  status: OperationalIncidentStatus
  source_domain: string
  source_identity: string
  deduplication_key: string
  occurrence: number
  opened_at: string
  last_observed_at: string
  acknowledged_at: string | null
  resolved_at: string | null
  summary: string
  evidence: Record<string, unknown>
  notification_state: 'NOT_ELIGIBLE' | 'SUPPRESSED_BY_PREFERENCE' | 'PENDING' | 'DELIVERED' | 'FAILED'
  events: OperationalIncidentEvent[]
}

export interface ForwardOperations {
  status: string
  scheduler_status: string
  scheduler_running: boolean
  cash: string | null
  equity: string | null
  open_positions: number
  pending_sessions: number
  latest_processed_session: string | null
  latest_completed_market_session: string | null
}

export interface BrokerOperations {
  enabled: boolean
  configured: boolean
  environment: string
  status: string
  last_success_at: string | null
  data_age_seconds: number | null
  snapshot_authoritative: boolean
}

export interface PositionDrift {
  ticker: string
  forward_quantity: string
  broker_quantity: string
  difference: string
  evaluated_at: string
}

export interface OperationsHealthResponse {
  overall_health: OperationalHealth
  critical_count: number
  warning_count: number
  info_count: number
  evaluated_at: string
  monitor_running: boolean
  monitor_status: string
  active_incidents: OperationalIncident[]
  forward: ForwardOperations
  broker: BrokerOperations
  position_drifts: PositionDrift[]
  startup: {
    database_reachable: boolean
    schema_compatible: boolean
    expected_schema_revision: string
    observed_schema_revision: string | null
    forward_scheduler_initialized: boolean
    broker_configuration_state: string
    operations_monitor_initialized: boolean
  }
}

export interface OperationsEvaluation {
  evaluated_at: string
  opened: number
  updated: number
  resolved: number
  overall_health: OperationalHealth
}

export type NotificationStatus = 'PENDING' | 'DELIVERING' | 'DELIVERED' | 'RETRY_PENDING' | 'FAILED' | 'CANCELLED'

export interface OperationalNotification {
  id: string
  incident_id: string | null
  channel: 'EMAIL'
  status: NotificationStatus
  kind: 'INCIDENT' | 'REMINDER' | 'RECOVERY' | 'DAILY_SUMMARY' | 'TEST'
  priority: 'INFO' | 'WARNING' | 'CRITICAL' | 'RECOVERED' | 'SUMMARY' | 'TEST'
  transition: string
  generation: number
  recipient: string
  subject: string
  deduplication_key: string
  trading_session: string | null
  scheduled_at: string
  next_attempt_at: string
  sent_at: string | null
  last_attempt_at: string | null
  attempt_count: number
  failure_category: string | null
  provider_message_reference: string | null
  created_at: string
  updated_at: string
  attempts: Array<{
    id: string
    attempt_number: number
    started_at: string
    completed_at: string
    result: 'DELIVERED' | 'TRANSIENT_FAILURE' | 'PERMANENT_FAILURE'
    failure_category: string | null
    provider_message_reference: string | null
    duration_ms: number
  }>
}

export interface NotificationDeliveryStatus {
  enabled: boolean
  email_enabled: boolean
  configured: boolean
  worker_running: boolean
  queue_depth: number
  pending_count: number
  failed_count: number
  last_delivery_at: string | null
  last_error: string | null
}

export interface NotificationPreferences {
  notifications_enabled: boolean
  email_enabled: boolean
  recipient: string | null
  warning_enabled: boolean
  critical_enabled: boolean
  recovery_enabled: boolean
  daily_summary_enabled: boolean
}

File: backend/src/alphapilot/schemas/operations.py
"""Public contracts for the deterministic Operations Center."""

from __future__ import annotations

from datetime import date
from decimal import Decimal
from typing import Any
from uuid import UUID

from pydantic import AwareDatetime, BaseModel, ConfigDict, Field

from alphapilot.database.models.operations import (
    OperationalEventType,
    OperationalHealth,
    OperationalIncidentStatus,
    OperationalIncidentType,
    OperationalSeverity,
    OperationalSourceDomain,
)


class OperationalIncidentEventSchema(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    event_type: OperationalEventType
    from_status: OperationalIncidentStatus | None
    to_status: OperationalIncidentStatus
    source: str
    reason: str
    evidence: dict[str, Any]
    created_at: AwareDatetime


class OperationalIncidentSchema(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    incident_type: OperationalIncidentType
    severity: OperationalSeverity
    status: OperationalIncidentStatus
    source_domain: OperationalSourceDomain
    source_identity: str
    deduplication_key: str
    occurrence: int
    opened_at: AwareDatetime
    last_observed_at: AwareDatetime
    acknowledged_at: AwareDatetime | None
    resolved_at: AwareDatetime | None
    summary: str
    evidence: dict[str, Any]
    notification_state: str = "NOT_ELIGIBLE"
    events: list[OperationalIncidentEventSchema] = Field(default_factory=list)


class StartupCheckSchema(BaseModel):
    database_reachable: bool
    schema_compatible: bool
    expected_schema_revision: str
    observed_schema_revision: str | None
    forward_scheduler_initialized: bool
    broker_configuration_state: str
    operations_monitor_initialized: bool


class ForwardOperationsSchema(BaseModel):
    status: str
    scheduler_status: str
    scheduler_running: bool
    cash: Decimal | None
    equity: Decimal | None
    open_positions: int
    pending_sessions: int
    latest_processed_session: date | None
    latest_completed_market_session: date | None


class BrokerOperationsSchema(BaseModel):
    enabled: bool
    configured: bool
    environment: str
    status: str
    last_success_at: AwareDatetime | None
    data_age_seconds: int | None
    snapshot_authoritative: bool


class PositionDriftSchema(BaseModel):
    ticker: str
    forward_quantity: Decimal
    broker_quantity: Decimal
    difference: Decimal
    evaluated_at: AwareDatetime


class OperationsHealthSchema(BaseModel):
    overall_health: OperationalHealth
    critical_count: int
    warning_count: int
    info_count: int
    evaluated_at: AwareDatetime
    monitor_running: bool
    monitor_status: str
    active_incidents: list[OperationalIncidentSchema]
    forward: ForwardOperationsSchema
    broker: BrokerOperationsSchema
    position_drifts: list[PositionDriftSchema]
    startup: StartupCheckSchema


class DailyOperationsSummarySchema(BaseModel):
    generated_at: AwareDatetime
    overall_health: OperationalHealth
    critical_count: int
    warning_count: int
    forward: ForwardOperationsSchema
    expected_manual_buys: int
    expected_manual_sells: int
    overdue_actions: int
    broker: BrokerOperationsSchema
    unmatched_broker_activity: int
    ambiguous_matches: int
    execution_conflicts: int
    position_drift_count: int
    latest_completed_market_session: date | None
    latest_forward_processed_session: date | None


class OperationsEvaluationSchema(BaseModel):
    evaluated_at: AwareDatetime
    opened: int
    updated: int
    resolved: int
    overall_health: OperationalHealth


class IncidentAcknowledgeRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    reason: str = Field(min_length=3, max_length=200)
